File: blackwatch-sim-convex/agents/agent.mjs
// Host diagnostics agent, Convex edition.
//
// The gRPC stream + reconnect/backlog machinery from the AWS version is gone:
// the agent subscribes to two reactive queries —
//   - mitigations.active : the full desired config set (push pipeline)
//   - hosts.me           : its own host doc (command channel; a status flip
//                          to "draining" arrives as a push)
// and reports health with a plain mutation. Resync-after-reboot is free:
// the subscription's current value IS the desired state.
import { ConvexClient } from "convex/browser";
import { api } from "../convex/_generated/api.js";

const args = Object.fromEntries(
  process.argv.slice(2).map((a) => a.replace(/^--/, "").split("=")),
);
const HOST_ID = args["host-id"];
const POP = args.pop ?? "iad-edge-1";
const LEAK = "leak" in args;
const URL = process.env.CONVEX_URL ?? "http://127.0.0.1:3310";
const POOL = 8 * 1024 ** 3;

const log = (m) => console.log(`[${HOST_ID}] ${m}`);

const dp = { mitigations: new Map(), wasted: 0 };
const used = () => [...dp.mitigations.values()].reduce((a, b) => a + b, 0) + dp.wasted;
const largestFreeBlock = () =>
  Math.floor(Math.max(POOL - used(), 0) / (1 + dp.mitigations.size / 2));

let lastLoadMs = 0;
let rebooting = false;
let desired = [];
let leaking = LEAK; // a fresh boot clears the fragmentation bug
let resyncing = false;

const client = new ConvexClient(URL);

function loadMitigation(m) {
  const start = Date.now();
  const cost = (200 + Math.floor(Math.random() * 200)) * 1024 ** 2;
  dp.mitigations.set(m.mitigationId, cost);
  if (leaking) dp.wasted += (2000 + Math.floor(Math.random() * 600)) * 1024 ** 2;
  lastLoadMs = Date.now() - start;
  const latency = resyncing
    ? "resync"
    : `propagation=${Date.now() - m.placedAt}ms`;
  log(
    `loaded ${m.mitigationId} action=${m.action} cidr=${m.targetCidr} ` +
      `${latency} largest_free_block=${Math.floor(largestFreeBlock() / 1024 ** 2)}MB`,
  );
}

function syncConfigs(mitigations) {
  desired = mitigations;
  if (rebooting) return;
  for (const m of mitigations) {
    if (!dp.mitigations.has(m.mitigationId)) loadMitigation(m);
  }
  for (const id of dp.mitigations.keys()) {
    if (!mitigations.some((m) => m.mitigationId === id)) {
      dp.mitigations.delete(id);
      log(`unloaded ${id}`);
    }
  }
}

async function reboot(reason) {
  rebooting = true;
  log(`REBOOT commanded (${reason}); draining 2s`);
  await new Promise((r) => setTimeout(r, 2000));
  dp.mitigations.clear();
  dp.wasted = 0;
  leaking = false;
  const epoch = await client.mutation(api.hosts.register, {
    hostId: HOST_ID,
    pop: POP,
    capacityGbps: 100,
  });
  rebooting = false;
  log(`rebooted, memory pool reset; re-registered epoch=${epoch}`);
  resyncing = true;
  syncConfigs(desired); // resync straight from the subscription's value
  resyncing = false;
}

const epoch = await client.mutation(api.hosts.register, {
  hostId: HOST_ID,
  pop: POP,
  capacityGbps: 100,
});
log(`registered, epoch=${epoch}`);

client.onUpdate(api.mitigations.active, {}, syncConfigs);
client.onUpdate(api.hosts.me, { hostId: HOST_ID }, (host) => {
  if (host?.status === "draining" && !rebooting) {
    void reboot("drain requested by orchestrator");
  }
});

setInterval(() => {
  if (rebooting) return;
  void client
    .mutation(api.hosts.heartbeat, {
      hostId: HOST_ID,
      cpuPct: 20 + Math.random() * 40,
      memoryUsedBytes: used(),
      largestFreeBlockBytes: largestFreeBlock(),
      activeMitigations: dp.mitigations.size,
      lastConfigLoadMs: lastLoadMs,
    })
    .catch((e) => log(`heartbeat failed: ${e.message}`));
}, 2000);

File: blackwatch-sim-aws/agent/agent.py
"""Host diagnostics agent for the scrubbing-fleet simulation.

Runs on each (simulated) scrubbing host:
  - registers with the control plane and opens the mitigation push stream
  - "loads" each mitigation into a simulated dataplane memory pool, measuring
    load time against the configured load SLA
  - reports health every HEALTH_INTERVAL_S, including the fragmentation
    early-warning metric `largest_free_block_bytes`
  - honors reboot/drain commands: stops taking configs, drains, re-registers

Simulated hosts never touch real traffic; the "dataplane" is a dict.
"""

import argparse
import random
import sys
import threading
import time

import grpc

sys.path.insert(0, __file__.rsplit("/", 2)[0] + "/gen/python")
import bwsim_pb2 as pb
import bwsim_pb2_grpc as rpc

HEALTH_INTERVAL_S = 2
LOAD_SLA_MS = 30_000
POOL_BYTES = 8 * 1024**3  # simulated dataplane memory pool


class Dataplane:
    """Toy model of mitigation filter memory (think: bloom filters, IPSets)."""

    def __init__(self, leak: bool):
        self.mitigations: dict[str, int] = {}  # id -> bytes allocated
        self.leak = leak
        self.wasted = 0  # bytes lost to fragmentation, grows on leaky hosts

    def load(self, m: pb.MitigationConfig) -> int:
        start = time.monotonic()
        # Filter structures cost 200-400MB depending on rule complexity.
        cost = random.randint(200, 400) * 1024**2
        self.mitigations[m.mitigation_id] = cost
        if self.leak:
            self.wasted += random.randint(2000, 2600) * 1024**2
        time.sleep(random.uniform(0.05, 0.25))  # simulated load work
        return int((time.monotonic() - start) * 1000)

    def used(self) -> int:
        return sum(self.mitigations.values()) + self.wasted

    def largest_free_block(self) -> int:
        free = max(POOL_BYTES - self.used(), 0)
        # Fragmentation model: contiguity degrades as more filters are packed in.
        return int(free / (1 + len(self.mitigations) / 2))


class Agent:
    def __init__(self, host_id: str, pop: str, target: str, leak: bool):
        self.host_id = host_id
        self.pop = pop
        self.target = target
        self.dp = Dataplane(leak)
        self.last_load_ms = 0
        self.running = True
        self.rebooting = threading.Event()

    def log(self, msg: str):
        print(f"[{self.host_id}] {msg}", flush=True)

    def run(self):
        while self.running:
            try:
                self.session()
            except grpc.RpcError as e:
                self.log(f"stream lost ({e.code().name}), reconnecting in 1s")
                time.sleep(1)

    def session(self):
        with grpc.insecure_channel(self.target) as channel:
            stub = rpc.MitigationServiceStub(channel)
            ack = stub.RegisterHost(
                pb.HostInfo(host_id=self.host_id, pop=self.pop, capacity_gbps=100)
            )
            self.log(f"registered, epoch={ack.stream_epoch}")
            self.rebooting.clear()

            reporter = threading.Thread(
                target=self.report_health, args=(stub,), daemon=True
            )
            reporter.start()

            for msg in stub.StreamMitigations(pb.StreamRequest(host_id=self.host_id)):
                if msg.HasField("mitigation"):
                    m = msg.mitigation
                    self.last_load_ms = self.dp.load(m)
                    sla = "OK" if self.last_load_ms <= LOAD_SLA_MS else "VIOLATION"
                    self.log(
                        f"loaded {m.mitigation_id} action={pb.Action.Name(m.action)} "
                        f"cidr={m.target_cidr} in {self.last_load_ms}ms (SLA {sla}) "
                        f"largest_free_block={self.dp.largest_free_block() // 1024**2}MB"
                    )
                elif msg.HasField("reboot"):
                    self.reboot(msg.reboot)
                    return

    def reboot(self, cmd: pb.RebootCommand):
        self.rebooting.set()
        self.log(f"REBOOT commanded (reason={cmd.reason!r}); draining {cmd.drain_seconds}s")
        time.sleep(cmd.drain_seconds)
        self.dp = Dataplane(leak=False)  # fresh boot clears fragmentation
        self.log("rebooted, memory pool reset; re-registering")

    def report_health(self, stub):
        while not self.rebooting.is_set():
            try:
                stub.ReportHealth(
                    pb.HealthReport(
                        host_id=self.host_id,
                        cpu_pct=random.uniform(20, 60),
                        memory_used_bytes=self.dp.used(),
                        largest_free_block_bytes=self.dp.largest_free_block(),
                        active_mitigations=len(self.dp.mitigations),
                        last_config_load_ms=self.last_load_ms,
                        reported_at_unix_ms=int(time.time() * 1000),
                    )
                )
            except grpc.RpcError:
                return
            time.sleep(HEALTH_INTERVAL_S)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--host-id", required=True)
    ap.add_argument("--pop", default="iad-edge-1")
    ap.add_argument("--target", default="localhost:50061")
    ap.add_argument(
        "--leak",
        action="store_true",
        help="simulate dataplane memory fragmentation so the orchestrator reboots us",
    )
    args = ap.parse_args()
    Agent(args.host_id, args.pop, args.target, args.leak).run()


if __name__ == "__main__":
    main()
